mesonbuild/build.py

```diff
@@ -970,7 +1001,15 @@ def link_whole(self, target):
                 raise InvalidArguments(msg)
             if self.is_cross != t.is_cross:
                 raise InvalidArguments('Tried to mix cross built and native libraries in target {!r}'.format(self.name))
-            self.link_whole_targets.append(t)
+            # When we're a static library and we link_whole: to another static
+            # library, we need to add that target's objects to ourselves.
+            if isinstance(self, StaticLibrary):
+                self.objects.append(t.extract_all_objects())
+                # Add internal and external deps
+                self.external_deps += t.external_deps
+                self.link_targets += t.link_targets
```

Comment:
Same here, we could need to recurse I think.

Reply:
Revisiting a bit this PR, and I have a question here: Why do you need to `extract_all_objects()` in the link_whole case? Isn't that exactly what the linker is going to do when we'll do `--whole-archive` ? Also since you removed `self.link_whole_targets.append(t)`, as far as I understand `self.link_whole_targets` list will now always be empty.

Probably I'm missing something here, but it sounds like we're never going to use --whole-archive any more, and always just give all objects files instead. Is that on purpose?

Reply:
Oh, sorry, didn't see the 'else' case, so `link_whole_targets` is still used for shared libraries.